minimint/src/net/connect.rs

```diff
@@ -1,175 +1,255 @@
-
-        let identity = &cfg.identity;
-        let handshakes = out_conns
-            .into_iter()
-            .chain(in_conns)
-            .map(move |mut stream| async move {
-                stream.write_u16((*identity).into()).await?;
-                let peer = stream.read_u16().await?.into();
-                Result::<_, std::io::Error>::Ok((peer, stream))
-            });
+    async fn connect_framed(&self, destination: String) -> ConnectResult<M> {
+        let mut connection = TcpStream::connect(destination).await?;
+        let peer = do_handshake(self.our_id, &mut connection).await?;
+        Ok((peer, Box::new(BidiFramed::new_from_tcp(connection))))
+    }
 
-        let peers = try_join_all(handshakes)
-            .await
-            .expect("Error during peer handshakes")
-            .into_iter()
-            .map(|(id, stream)| (id, Framed::new(stream.compat())))
-            .collect::<HashMap<_, _>>();
+    async fn listen(
+        &self,
+        bind_addr: String,
+    ) -> Result<Pin<Box<dyn Stream<Item = ConnectResult<M>> + Send + Unpin + 'static>>, anyhow::Error>
+    {
+        let listener = TcpListener::bind(bind_addr).await?;
+        let our_id = self.our_id;
+
+        let stream = futures::stream::unfold(listener, move |listener| {
+            Box::pin(async move {
+                let res = listener.accept().await;
+
+                let res = match res {
+                    Ok((mut connection, _)) => do_handshake(our_id, &mut connection)
+                        .await
+                        .map(|peer| (peer, BidiFramed::new_from_tcp(connection).to_any())),
+                    Err(e) => Err(anyhow::Error::new(e)),
+                };
+
+                Some((res, listener))
+            })
+        });
+        Ok(Box::pin(stream))
+    }
+}
 
-        info!("Successfully connected to all peers");
+async fn do_handshake<S>(our_id: PeerId, stream: &mut S) -> Result<PeerId, anyhow::Error>
+where
+    S: AsyncRead + AsyncWrite + Unpin,
+{
+    // Send our id
+    let our_id = our_id.to_usize() as u16;
+    stream.write_all(&our_id.to_le_bytes()[..]).await?;
+
+    // Receive peer id
+    let mut peer_id = [0u8; 2];
+    stream.read_exact(&mut peer_id[..]).await?;
+    Ok(PeerId::from(u16::from_le_bytes(peer_id)))
+}
 
-        Connections { connections: peers }
+#[allow(unused_imports)]
+pub mod mock {
+    use crate::net::connect::{do_handshake, ConnectResult, Connector};
+    use crate::net::framed::{BidiFramed, FramedTransport};
+    use anyhow::Error;
+    use futures::{FutureExt, SinkExt, Stream, StreamExt};
+    use minimint_api::PeerId;
+    use std::collections::HashMap;
+    use std::fmt::Debug;
+    use std::future::Future;
+    use std::pin::Pin;
+    use std::sync::Arc;
+    use std::time::Duration;
+    use tokio::io::{DuplexStream, ReadHalf, WriteHalf};
+    use tokio::sync::mpsc::Sender;
+    use tokio::sync::Mutex;
+
+    pub struct MockNetwork {
```

Comment:
Is the advantage of using mocks that we don't need to depend on ports being open?